OIDC: Encapsulate static/dynamic tenants maps in `TenantConfigBean`

snazy opened the pull request:
Lets `TenantConfigBean` be the sole "owner" of the static/dynamic tenants maps, adds/changes accessor methods for tenants. Also introduces a functional interface to create tenants.

No functional change, only moving code around.

Also some minor code cleanups (reducing the amount of warnings in IntelliJ).

sberyozkin commented:
@snazy Thanks, can you fix the formatting please, I'll have a look tomorrow once I finish another PR

michalvavrik commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java:
```diff
@@ -43,6 +43,18 @@ public class OidcConfig {
+    @ConfigGroup
+    public static class DynamicTenants {
+        /**
+         * If configured, limit the number of active OID tenant contexts to this value.
```
```suggestion
         * If configured, limit the number of active OIDC tenant contexts to this value.
```

michalvavrik commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcConfig.java:
```diff
@@ -43,6 +43,18 @@ public class OidcConfig {
+         * If configured, limit the number of active OID tenant contexts to this value.
+         */
+        @ConfigItem
+        public int limit;
```
I don't see this property used anywhere, yet, it is part of the config now.

snazy replied:
Yea - removed the config change

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTenantConfigResolver.java:
```diff
@@ -119,10 +117,10 @@ public OidcTenantConfig apply(OidcTenantConfig tenantConfig) {
 
                             if (tenantId != null && !isTenantSetByAnnotation(context, tenantId)) {
-                                TenantConfigContext tenantContext = tenantConfigBean.getDynamicTenantsConfig().get(tenantId);
+                                TenantConfigContext tenantContext = tenantConfigBean.getExistingTenantConfigContext(tenantId);
```
`Existing` includes both static and dynamic ones, so propose to keep the original name for now

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java:
```diff
@@ -1,66 +1,115 @@
-        return dynamicTenantsConfig;
+    public Uni<TenantConfigContext> getOrCreateTenantContext(OidcTenantConfig oidcConfig, boolean dynamicTenant) {
+        // check if the connection has already been created
+        Map<String, TenantConfigContext> tenantsConfig = dynamicTenant ? dynamicTenantsConfig : staticTenantsConfig;
```
For dynamic tenants, the connection is established during the request, so AFAIK, their context always must be in the ready state if it is in recorded in the map

sberyozkin commented:
@snazy Thanks, it all looks good, refactoring makes sense and makes things a bit clearer. Michal, myself, added minor comments, but something did not work out during the initial refactoring, tests are impacted

snazy commented:
The issue that's fixed now is that `TenantConfigContext`s for static tenants with a `null` `provider` were "leaked", causing the NPEs. Also updated the code to make things a bit clearer/leaner.

snazy commented:
Um - let me look into those new test failures

snazy commented:
Hm - the tests in those projects don't fail locally

michalvavrik commented:
> Hm - the tests in those projects don't fail locally

Did you declare you have Linux containers available locally with `-Dtest-containers -Dstart-containers` appended to your `mvn clean ....`?

snazy commented:
> > Hm - the tests in those projects don't fail locally
> 
> Did you declare you have Linux containers available locally with `-Dtest-containers -Dstart-containers` appended to your `mvn clean ....`?

Yup - just found those params in the GH workflow

sberyozkin commented:
@snazy One thing I know is that even a single line update can have an impact there, a lot is going on with the tenant resolution, and this is why, having a bit of clean up and adding extra clarity would not be bad at all. It ll help going forward.
But I wonder, would it make sense to do a series of really small PRs for us to have a better control of these changes ? 
For example:
* PR1: get static tenants created at the first request time recorded in the static map
* PR2: Consolidate the code which is really belonging to the tenant bean in that bean

snazy commented:
Yea - something is quite broken with this change. Let me start from scratch.

sberyozkin commented:
Thanks @snazy, yeah, I think the introduction of the functional interface is nice, and making sure static tenants are only recorded in the static map, is worth a dedicated PR of its own, so if you are OK with it, please do it first, and then complete the preparation with the 2nd one

snazy commented:
Marked as a draft PR for now.

snazy committed:
OIDC: Encapsulate static/dynamic tenants maps in `TenantConfigBean`

Lets `TenantConfigBean` be the sole "owner" of the static/dynamic tenants maps, adds/changes accessor methods for tenants. Also introduces a functional interface to create tenants.

No functional change, only moving code around.

Also some minor code cleanups (reducing the amount of warnings in IntelliJ).

snazy commented:
Ah! I found out why the PR fails in the `CodeFlowTest`: I've changed the order of the dynamic/static tenant checks in `i.q.oidc.runtime.DefaultTenantConfigResolver#resolveConfig`, which broke the tests. Which means, the refactoring was broken in the sense that static tenants still leaked into the dynamic map. Anyway - not a problem, as I'm going to change this PR.

snazy commented:
Updated the PR entirely, focusing on moving the tenants-maps into the `TenantConfigBean` and not let these maps escape.

snazy commented:
Note: `OidcClientRegistrationTest.testRegisteredClientDynamicTenant` in `integration-tests/oidc-client-registration/` fails on `main` as well.

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java:
```diff
@@ -1,46 +1,75 @@
-    public Map<String, TenantConfigContext> getStaticTenantsConfig() {
-        return staticTenantsConfig;
+    public Uni<TenantConfigContext> createTenantContext(OidcTenantConfig oidcConfig, boolean dynamicTenant) {
+        if (oidcConfig.logout.backchannel.path.isPresent()) {
```
`dynamicTenant` is not checked yet

snazy replied:
Static tenants still escape into the dynamicTenants map in this PR.
I've opened #43110 to let static tenants not escape - tests should pass on 43110 now - the actual change wasn't that big - could actually sneak it back into this PR. WDYT?

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTenantConfigResolver.java:
```diff
@@ -119,17 +119,18 @@ public OidcTenantConfig apply(OidcTenantConfig tenantConfig) {
                                 if (tenantContext != null) {
                                     // Dynamic map may contain the static contexts initialized on demand,
-                                    if (tenantConfigBean.getStaticTenantsConfig().containsKey(tenantId)) {
+                                    if (tenantConfigBean.getStaticTenant(tenantId) != null) {
```
Please remove the comment above since it is no longer relevant

snazy replied:
#43110 removes it (the comment's still valid for this PR)

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/TenantConfigBean.java:
```diff
@@ -1,46 +1,75 @@
-        this.staticTenantsConfig = staticTenantsConfig;
-        this.dynamicTenantsConfig = dynamicTenantsConfig;
+            TenantContextFactory tenantContextFactory) {
+        this.staticTenantsConfig = new ConcurrentHashMap<>(staticTenantsConfig);
```
It is not clear to me why it is now a concurrent map, given the `dynamicTenant` in the function below is not checked

snazy replied:
#43110 should make it clear

sberyozkin commented:
Hi @snazy Can you please pickup the latest updates and rework this PR next week or so, to have access to the dynamic (and static) map encapsulated inside `TenantConfigBean`, and replacing an explicit Function reference with an interface;

Now that the static tenant does not leak into the dynamic map anymore after a fix from @michalvavrik, the PR should become much simpler, 

Let me now please if you won't have much time, thanks.

snazy commented:
Closed in favor of #43590
